### PR TITLE
Science sensors software 2022 r1

### PR DESCRIPTION
**Pull request by antrob25**

### CODE REVIEW

File: SRA_Sensors/SRA_Sensors.ino [hunk 1]
```diff
@@ -1,60 +1,89 @@
...
+void setup()
+{
+  timing = 0;   //Initialize the timing variable
+
+  Computer_Serial.begin(115200);
+  CO2_Serial.begin(19200);
+  O2_Serial.begin(9600);  
+  CH4_Serial.begin(115200);
+  myusb.begin();
+  userial.begin(115200);
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: These baud rates should be put in consts

File: SRA_Sensors/SRA_Sensors.ino [hunk 2]
```diff
@@ -1,60 +1,89 @@
...
   }
-  timing++; //Increment the timer //TODO: Create constants for polling rates?
+  timing++;   //Increment the timer
   delay(100);
 }
 
 void updateLed(int msg)
 {
-  if(msg==RC_SCIENCESENSORSBOARD_UVLEDCONTROL_DATA_ID)
+  if(msg==RC_SCIENCESENSORSBOARD_LIGHTS_DATA_ID)
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: The Lights command is a bitmasked value where the 0th bit is whether the White light is on, and the 1th bit is whether the UV light is on. As it is now, this if statement will never eval to true.

File: SRA_Sensors/SRA_Sensors.ino [hunk 3]
```diff
@@ -1,60 +1,89 @@
...
         updateLed((int)packet.data[0]);
         break;
+      case RC_SCIENCESENSORSBOARD_FLASERS_DATA_ID:
+        uint8_t* lasersOn = (uint8_t*)packet.data;
+        if(lasersOn[0] != 0)
+        {
+          digitalWrite(Laser1, HIGH);
+          digitalWrite(Laser2, HIGH);
+          digitalWrite(Laser3, HIGH); 
+        }
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: The FLasers command is a bitmasked value which controls the lasers individually. How it is now, the lasers can only be controlled all at once.

File: SRA_Sensors/SRA_Sensors.ino [hunk 4]
```diff
@@ -1,60 +1,89 @@
...
     }
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: Should have a break statement for the FLASERS case and a default case

File: SRA_Sensors/SRA_Sensors.ino [hunk 5]
```diff
@@ -65,29 +94,66 @@ void updateLed(int msg)
...
+  uint16_t raw3 = analogRead(Photodiode3);
+
+  // get voltage values
+  float v1 = raw1 * ref_voltage / adc_resolution;
+  float v2 = raw2 * ref_voltage / adc_resolution;
+  float v3 = raw3 * ref_voltage / adc_resolution;
+
+  Serial.println((float)v1);
+  Serial.println((float)v2);
+  Serial.println((float)v3);
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: Don't have to cast floats to floats here

File: SRA_Sensors/SRA_Sensors.ino [hunk 6]
```diff
@@ -65,29 +94,66 @@ void updateLed(int msg)
...
+  pdreadings[2] = reverseCurrent*(exp(-(50/3) * v3)-1) - darkCurrent;
+
+  RoveComm.write(RC_SCIENCESENSORSBOARD_FLUOROMETERDATA_DATA_ID,RC_SCIENCESENSORSBOARD_FLUOROMETERDATA_DATA_COUNT, pdreadings);
+}
+
+void o2Reading()
+{
+    if(O2_Serial.read()=='%') //Start at the beginning of the o2 Percentage sensor output
+    {
+      float o2readings[1];
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: Doesn't have to be an array

File: SRA_Sensors/SRA_Sensors.ino [hunk 7]
```diff
@@ -65,29 +94,66 @@ void updateLed(int msg)
...
 }
 
-void co2Reading(){
+void ch4Reading()
+{
+    if(userial.read() =='C') //Start at the beginning of the ch4 sensor output
+    {
+      if(userial.read() ==':') //Start at the beginning of the ch4 sensor output
+      {
+        userial.read();
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: (Repeat this for all lines like this one) Maybe add a comment to explain why you're throwing away a byte. I know it's to skip a space, but should be clear about it.

File: SRA_Sensors/SRA_Sensors.ino [hunk 8]
```diff
@@ -1,60 +1,89 @@
...
+        {
+          digitalWrite(Laser1, HIGH);
+          digitalWrite(Laser2, HIGH);
+          digitalWrite(Laser3, HIGH); 
+        }
     }
   }
-  timing++; //Increment the timer //TODO: Create constants for polling rates?
+  timing++;   //Increment the timer
   delay(100);
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.ino]: Needs proper timing/interrupt logic

**Review by ThisIsBrady — CHANGES_REQUESTED** [review]

File: SRA_Sensors/SRA_Sensors.h [hunk 9]
```diff
@@ -1,18 +1,52 @@
...
+#define ADC_CLK 1600000
+#define ref_voltage 5
+#define adc_resolution 4096
+
+#define Photodiode1 A11
+#define Photodiode2 A10
+#define Photodiode3 A9
+
+#define reverseCurrent pow(10,-15)
+#define darkCurrent 0 //this is wrong but I don't think there's a way to get this number without measurement
```
Comment (ThisIsBrady) [inline SRA_Sensors/SRA_Sensors.h]: What's dark current?

File: SRA_Sensors/SRA_Sensors.ino [hunk 10]
```diff
@@ -65,29 +94,66 @@ void updateLed(int msg)
...
+  pdreadings[0] = reverseCurrent*(exp(-(50/3) * v1)-1) - darkCurrent; 
+  pdreadings[1] = reverseCurrent*(exp(-(50/3) * v2)-1) - darkCurrent;
+  pdreadings[2] = reverseCurrent*(exp(-(50/3) * v3)-1) - darkCurrent;
+
+  RoveComm.write(RC_SCIENCESENSORSBOARD_FLUOROMETERDATA_DATA_ID,RC_SCIENCESENSORSBOARD_FLUOROMETERDATA_DATA_COUNT, pdreadings);
+}
+
+void o2Reading()
+{
+    if(O2_Serial.read()=='%') //Start at the beginning of the o2 Percentage sensor output
```
Comment (EliVerbrugge) [inline SRA_Sensors/SRA_Sensors.ino]: I'd recommend dropping a quick comment to describe what the serial packet looks like, e.g. how you know to drop %

**Pull request closed without merging**